NullPointerException in UniqueNameHelper.getValidDN

Calls to UniqueNameHelper.getValidDN can result in NullPointerExceptions when testing whether a null String is a valid DN.

An exception stack similar to that below may be found in the trace or FFDC:

```
Caused by: java.lang.NullPointerException
	at javax.naming.ldap.Rfc2253Parser.<init>(Rfc2253Parser.java:59)
	at javax.naming.ldap.LdapName.parse(LdapName.java:796)
	at javax.naming.ldap.LdapName.<init>(LdapName.java:134)
	at com.ibm.ws.security.wim.util.UniqueNameHelper.getValidDN(UniqueNameHelper.java:99)
	at com.ibm.ws.security.wim.util.UniqueNameHelper.isDNUnderBaseEntry(UniqueNameHelper.java:107)
	at com.ibm.ws.security.wim.adapter.ldap.LdapHelper.isUnderBases(LdapHelper.java:364)
	at com.ibm.ws.security.wim.adapter.ldap.LdapAdapter.getMembersByMember(LdapAdapter.java:2206)
	at com.ibm.ws.security.wim.adapter.ldap.LdapAdapter.getMembers(LdapAdapter.java:1930)
	at com.ibm.ws.security.wim.adapter.ldap.LdapAdapter.get(LdapAdapter.java:452)
	at com.ibm.ws.security.wim.ProfileManager.getImpl(ProfileManager.java:545)
	... 41 more
```

--- a/dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/FATTestIDS_allIbmGroups.java
+++ b/dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/FATTestIDS_allIbmGroups.java
@@ -185,7 +185,7 @@ public class FATTestIDS_allIbmGroups {
              * Configure group membership.
              */
             GroupProperties groupProperties = new GroupProperties();
-            groupProperties.setMemberAttribute(new MemberAttribute(null, "ibm-allmembers", "groupofnames", "all"));
+            groupProperties.getMemberAttributes().add(new MemberAttribute(null, "ibm-allmembers", "groupofnames", "all"));
             groupProperties.setMembershipAttribute(new MembershipAttribute("ibm-allgroups", "all"));
             ldap.setGroupProperties(groupProperties);
         }

--- a/dev/com.ibm.ws.security.wim.core/src/com/ibm/ws/security/wim/util/UniqueNameHelper.java
+++ b/dev/com.ibm.ws.security.wim.core/src/com/ibm/ws/security/wim/util/UniqueNameHelper.java
@@ -42,7 +42,7 @@ public class UniqueNameHelper {
         if (uniqueName == null) {
             return null;
         }
-        return getValidDN(uniqueName);
+        return getValidUniqueName(uniqueName);
     }
 
     /**
@@ -80,12 +80,13 @@ public class UniqueNameHelper {
      * @return The valid and formatted unique name. null will be returned if the unique name is invalid.
      */
     @Trivial
-    public static String getValidUniqueName(String uniqueName) {
-        return getValidDN(uniqueName);
-    }
-
     @FFDCIgnore(InvalidNameException.class)
-    private static String getValidDN(String uniqueName) {
+    public static String getValidUniqueName(String uniqueName) {
+
+        if (uniqueName == null) {
+            return null;
+        }
+
         try {
             /*
              * This may seem excessively complex. Well it is. The issue is that the
@@ -104,7 +105,7 @@ public class UniqueNameHelper {
     }
 
     public static boolean isDNUnderBaseEntry(String dn, String... bases) {
-        dn = getValidDN(dn);
+        dn = getValidUniqueName(dn);
         if (dn != null) {
             dn = dn.toLowerCase();
             //return true if any of bases is empty node/""/root or if dn ends with any of the the base

--- a/dev/com.ibm.ws.security.wim.core/test/com/ibm/ws/security/wim/util/UniqueNameHelperTest.java
+++ b/dev/com.ibm.ws.security.wim.core/test/com/ibm/ws/security/wim/util/UniqueNameHelperTest.java
@@ -55,6 +55,7 @@ public class UniqueNameHelperTest {
         /*
          * Test some of the space handling in getValidDN.
          */
+        assertNull(UniqueNameHelper.getValidUniqueName(null));
         assertEquals("uid=user1,dc=ibm,dc=com", UniqueNameHelper.getValidUniqueName("uid=user1,dc=ibm,dc=com"));
         assertEquals("uid=user1 \\ ,dc=ibm,dc=com", UniqueNameHelper.getValidUniqueName("uid=user1 \\ ,dc=ibm,dc=com"));
         assertEquals("uid=user1    \\ ,dc=ibm,dc=com", UniqueNameHelper.getValidUniqueName("uid=user1\\ \\ \\ \\ \\ ,dc=ibm,dc=com"));
@@ -69,6 +70,7 @@ public class UniqueNameHelperTest {
         /*
          * Test some of the space handling in getValidDN.
          */
+        assertNull(UniqueNameHelper.isDN(null));
         assertEquals("uid=user1,dc=ibm,dc=com", UniqueNameHelper.isDN("uid=user1,dc=ibm,dc=com"));
         assertEquals("uid=user1 \\ ,dc=ibm,dc=com", UniqueNameHelper.isDN("uid=user1 \\ ,dc=ibm,dc=com"));
         assertEquals("uid=user1    \\ ,dc=ibm,dc=com", UniqueNameHelper.isDN("uid=user1\\ \\ \\ \\ \\ ,dc=ibm,dc=com"));

--- a/dev/com.ibm.ws.security.wim.core_fat/bnd.bnd
+++ b/dev/com.ibm.ws.security.wim.core_fat/bnd.bnd
@@ -27,4 +27,6 @@ fat.project: true
 	com.ibm.ws.security.registry;version=latest,\
 	com.ibm.ws.security.registry_test.servlet;version=latest,\
 	com.ibm.ws.security.wim.core;version=latest,\
-	com.ibm.wsspi.org.osgi.service.component.annotations;version=latest
+	com.ibm.wsspi.org.osgi.service.component.annotations;version=latest,\
+	com.ibm.ws.com.unboundid;version=latest,\
+	com.unboundid:unboundid-ldapsdk;version=latest

--- a/dev/com.ibm.ws.security.wim.core_fat/build.gradle
+++ b/dev/com.ibm.ws.security.wim.core_fat/build.gradle
@@ -12,7 +12,9 @@
 dependencies { 
   requiredLibs project(':fattest.simplicity'),
     project(':com.ibm.ws.security.registry'),
-    project(':com.ibm.ws.security.registry_test.servlet')
+    project(':com.ibm.ws.security.registry_test.servlet'),
+    'com.unboundid:unboundid-ldapsdk:4.0.9',
+    project(':com.ibm.ws.com.unboundid')
 }
 autoFVT.dependsOn ':com.ibm.ws.security.registry_test.servlet:war'
 autoFVT.doLast {
@@ -33,7 +35,8 @@ autoFVT.doLast {
     'com.ibm.ws.security.wim.core.fat.maxSearchResult',
     'com.ibm.ws.security.wim.core.fat.searchTimeOut',
     'com.ibm.ws.security.wim.core.fat.dynamic',
-    'com.ibm.ws.security.wim.core.fat.noreg'
+    'com.ibm.ws.security.wim.core.fat.noreg',
+    'com.ibm.ws.security.wim.core.fat.WimCoreRegression'
   ]
   servers.each { server ->
     copy {

--- a/dev/com.ibm.ws.security.wim.core_fat/fat/src/com/ibm/ws/security/wim/core/fat/FATSuite.java
+++ b/dev/com.ibm.ws.security.wim.core_fat/fat/src/com/ibm/ws/security/wim/core/fat/FATSuite.java
@@ -19,7 +19,8 @@ import org.junit.runners.Suite.SuiteClasses;
 @SuiteClasses({ InvalidBaseEntryInRealmTest.class,
                 MaxSearchResultTest.class,
                 DynamicUpdateTest.class,
-                NoRegistryConfiguredTest.class
+                NoRegistryConfiguredTest.class,
+                WimCoreRegressionTest.class
 })
 /**
  * Purpose: This suite collects and runs all known good test suites.

--- a/dev/com.ibm.ws.security.wim.core_fat/fat/src/com/ibm/ws/security/wim/core/fat/WimCoreRegressionTest.java
+++ b/dev/com.ibm.ws.security.wim.core_fat/fat/src/com/ibm/ws/security/wim/core/fat/WimCoreRegressionTest.java
@@ -1,0 +1,241 @@
+/*******************************************************************************
+ * Copyright (c) 2019 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+
+package com.ibm.ws.security.wim.core.fat;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+import java.util.Arrays;
+
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import com.ibm.websphere.simplicity.config.ServerConfiguration;
+import com.ibm.websphere.simplicity.config.wim.GroupProperties;
+import com.ibm.websphere.simplicity.config.wim.LdapRegistry;
+import com.ibm.websphere.simplicity.config.wim.MemberAttribute;
+import com.ibm.websphere.simplicity.log.Log;
+import com.ibm.ws.com.unboundid.InMemoryLDAPServer;
+import com.ibm.ws.security.registry.SearchResult;
+import com.ibm.ws.security.registry.test.UserRegistryServletConnection;
+import com.unboundid.ldap.sdk.Entry;
+
+import componenttest.custom.junit.runner.FATRunner;
+import componenttest.custom.junit.runner.Mode;
+import componenttest.custom.junit.runner.Mode.TestMode;
+import componenttest.topology.impl.LibertyServer;
+import componenttest.topology.impl.LibertyServerFactory;
+import componenttest.topology.utils.LDAPFatUtils;
+import componenttest.topology.utils.LDAPUtils;
+
+/**
+ * Some regression tests for the com.ibm.ws.security.wim.core component.
+ */
+@RunWith(FATRunner.class)
+@Mode(TestMode.LITE)
+public class WimCoreRegressionTest {
+    private static LibertyServer libertyServer = LibertyServerFactory.getLibertyServer("com.ibm.ws.security.wim.core.fat.WimCoreRegression");
+    private static final Class<?> c = WimCoreRegressionTest.class;
+    private static UserRegistryServletConnection servlet;
+    private static InMemoryLDAPServer ds = null;
+    private static ServerConfiguration startConfiguration = null;
+
+    private static final String BASE_DN = "o=ibm.com";
+    private static final String GROUP_CN = "group1";
+    private static final String GROUP_DN = "cn=" + GROUP_CN + "," + BASE_DN;
+
+    private static final String USER_UID = "user1";
+    private static final String USER_DN = "uid=" + USER_UID + "," + BASE_DN;
+    private static final String USER_PASSWORD = "password";
+
+    /**
+     * Setup the test case.
+     *
+     * @throws Exception If the setup failed for some reason.
+     */
+    @BeforeClass
+    public static void setupClass() throws Exception {
+        setupLibertyServer();
+        setupLdapServer();
+    }
+
+    /**
+     * Tear down the test.
+     */
+    @AfterClass
+    public static void teardownClass() throws Exception {
+        try {
+            if (libertyServer != null) {
+                libertyServer.stopServer();
+            }
+        } finally {
+            try {
+                if (ds != null) {
+                    ds.shutDown(true);
+                }
+            } catch (Exception e) {
+                Log.error(c, "teardown", e, "LDAP server threw error while shutting down. " + e.getMessage());
+            }
+        }
+    }
+
+    /**
+     * Setup the Liberty server. This server will start with very basic configuration. The tests
+     * will configure the server dynamically.
+     *
+     * @throws Exception If there was an issue setting up the Liberty server.
+     */
+    private static void setupLibertyServer() throws Exception {
+        /*
+         * Add LDAP variables to bootstrap properties file
+         */
+        LDAPUtils.addLDAPVariables(libertyServer);
+        Log.info(c, "setUpLibertyServer", "Starting the server... (will wait for userRegistry servlet to start)");
+        libertyServer.copyFileToLibertyInstallRoot("lib/features", "internalfeatures/securitylibertyinternals-1.0.mf");
+        libertyServer.addInstalledAppForValidation("userRegistry");
+        libertyServer.startServer(c.getName() + ".log");
+
+        /*
+         * Make sure the application has come up before proceeding
+         */
+        assertNotNull("Application userRegistry does not appear to have started.",
+                      libertyServer.waitForStringInLog("CWWKZ0001I:.*userRegistry"));
+        assertNotNull("Security service did not report it was ready",
+                      libertyServer.waitForStringInLog("CWWKS0008I"));
+        assertNotNull("Server did not came up",
+                      libertyServer.waitForStringInLog("CWWKF0011I"));
+
+        Log.info(c, "setUp", "Creating servlet connection the server");
+        servlet = new UserRegistryServletConnection(libertyServer.getHostname(), libertyServer.getHttpDefaultPort());
+
+        if (servlet.getRealm() == null) {
+            Thread.sleep(5000);
+            servlet.getRealm();
+        }
+
+        /*
+         * The original server configuration.
+         */
+        startConfiguration = libertyServer.getServerConfiguration();
+    }
+
+    /**
+     * Configure the in-memory LDAP server.
+     *
+     * @throws Exception If the server failed to start for some reason.
+     */
+    private static void setupLdapServer() throws Exception {
+        ds = new InMemoryLDAPServer(BASE_DN);
+        Log.info(c, "setUpldapServer", "populate LDAP Server");
+
+        /*
+         * Add the partition entries.
+         */
+        Entry entry = new Entry(BASE_DN);
+        entry.addAttribute("objectclass", "organization");
+        entry.addAttribute("o", "ibm.com");
+        ds.add(entry);
+
+        /*
+         * Create user.
+         */
+        entry = new Entry(USER_DN);
+        entry.addAttribute("objectclass", "inetorgperson");
+        entry.addAttribute("uid", USER_UID);
+        entry.addAttribute("sn", USER_UID + "_sn");
+        entry.addAttribute("cn", USER_UID + "_cn");
+        entry.addAttribute("userPassword", USER_PASSWORD);
+        ds.add(entry);
+
+        /*
+         * Create group.
+         */
+        entry = new Entry(GROUP_DN);
+        entry.addAttribute("objectclass", "wimGroupOfNames");
+        entry.addAttribute("cn", GROUP_CN);
+        entry.addAttribute("member", USER_DN);
+
+        ds.add(entry);
+
+    }
+
+    /**
+     * Create a basic LDAP registry configuration element for the specified server configuration.
+     * The registry will be added to the configuration.
+     *
+     * @param serverConfig The server configuration to add the LDAP registry to.
+     * @return The new LDAP registry configuration element.
+     */
+    private static LdapRegistry createLdapRegistry(ServerConfiguration serverConfig) {
+        /*
+         * Create and add the new LDAP registry to the server configuration.
+         */
+        LdapRegistry ldapRegistry = new LdapRegistry();
+        serverConfig.getLdapRegistries().add(ldapRegistry);
+
+        /*
+         * Configure the LDAP registry.
+         */
+        ldapRegistry.setBaseDN(BASE_DN);
+        ldapRegistry.setLdapType("Custom");
+        ldapRegistry.setRealm("LdapRealm");
+        ldapRegistry.setHost("localhost");
+        ldapRegistry.setPort(String.valueOf(ds.getListenPort()));
+        ldapRegistry.setBindDN(InMemoryLDAPServer.getBindDN());
+        ldapRegistry.setBindPassword(InMemoryLDAPServer.getBindPassword());
+        ldapRegistry.setTimestampFormat("yyyyMMddHHmmss.SSSSZ"); // 20180730202338.850-0000Z
+        GroupProperties groupProperties = new GroupProperties();
+        groupProperties.getMemberAttributes().add(new MemberAttribute(null, "member", "groupOfNames", "direct"));
+        groupProperties.getMemberAttributes().add(new MemberAttribute(null, "ibm-groupMember", "wimGroupOfNames", "direct"));
+        ldapRegistry.setGroupProperties(groupProperties);
+
+        return ldapRegistry;
+    }
+
+    /**
+     * Regression test for Open Liberty GitHub issue 8034.
+     *
+     * This particular case occurred due to the way that cached attributes are stored in the LDAP registry. The
+     * LDAP registry stores any attributes that are not found in the LDAP cache with a null value. Since the
+     * group being retrieved didn't have a value for 'ibm-groupMember', it was stored in the cache with a
+     * null value. On the next lookup, the search returned that null value when doing membership lookup
+     * and the UniqueNameHelper.getValidDN(...) call would error with an NPE when trying to parse the null
+     * as a DN.
+     *
+     * @throws Exception if the test failed for some unforeseen reason.
+     */
+    @Test
+    public void multiValuedOutputProperties() throws Exception {
+
+        ServerConfiguration clone = startConfiguration.clone();
+        createLdapRegistry(clone);
+        LDAPFatUtils.updateConfigDynamically(libertyServer, clone);
+
+        /*
+         * In the error case, the group lookup would succeed on the first call b/c the JNDI
+         * call to LDAP would not contain the attribute with the null value.
+         */
+        SearchResult result = servlet.getUsersForGroup(GROUP_DN, 0);
+        assertEquals("Expected user " + USER_DN + " to be a member of group " + GROUP_DN, Arrays.asList(new String[] { USER_DN }), result.getList());
+
+        /*
+         * The second call would fail with the NPE since the group membership lookup would
+         * get a response from the LDAP cache that contained the 'ibm-groupMember' attribute
+         * with a null value. The code would check to see if the contained value was a DN
+         * by calling UniqueNameHelper.getValidDN and that would fail with the NPE.
+         */
+        result = servlet.getUsersForGroup(GROUP_DN, 0);
+        assertEquals("Expected user " + USER_DN + " to be a member of group " + GROUP_DN, Arrays.asList(new String[] { USER_DN }), result.getList());
+    }
+}

--- a/dev/com.ibm.ws.security.wim.core_fat/publish/servers/com.ibm.ws.security.wim.core.fat.WimCoreRegression/bootstrap.properties
+++ b/dev/com.ibm.ws.security.wim.core_fat/publish/servers/com.ibm.ws.security.wim.core.fat.WimCoreRegression/bootstrap.properties
@@ -1,0 +1,3 @@
+com.ibm.ws.logging.trace.specification=*=info=enabled:com.ibm.ws.security.*=all=enabled
+com.ibm.ws.logging.max.file.size=0
+bootstrap.include=../testports.properties

--- a/dev/com.ibm.ws.security.wim.core_fat/publish/servers/com.ibm.ws.security.wim.core.fat.WimCoreRegression/server.xml
+++ b/dev/com.ibm.ws.security.wim.core_fat/publish/servers/com.ibm.ws.security.wim.core.fat.WimCoreRegression/server.xml
@@ -1,0 +1,10 @@
+<server description="WimCoreRegression">
+
+	<include location="../fatTestPorts.xml"/>
+    <!-- Enable features -->
+    <featureManager>
+        <feature>appSecurity-1.0</feature>
+        <feature>securitylibertyinternals-1.0</feature>
+        <feature>ldapRegistry-3.0</feature>
+	</featureManager>
+</server>

--- a/dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/wim/GroupProperties.java
+++ b/dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/wim/GroupProperties.java
@@ -13,6 +13,7 @@ package com.ibm.websphere.simplicity.config.wim;
 import javax.xml.bind.annotation.XmlElement;
 
 import com.ibm.websphere.simplicity.config.ConfigElement;
+import com.ibm.websphere.simplicity.config.ConfigElementList;
 
 /**
  * Configuration for the following nested elements:
@@ -24,7 +25,7 @@ import com.ibm.websphere.simplicity.config.ConfigElement;
 public class GroupProperties extends ConfigElement {
 
     private DynamicMemberAttribute dynamicMemberAttribute;
-    private MemberAttribute memberAttribute;
+    private ConfigElementList<MemberAttribute> memberAttribute;
     private MembershipAttribute membershipAttribute;
 
     /**
@@ -37,8 +38,8 @@ public class GroupProperties extends ConfigElement {
     /**
      * @return the memberAttribute
      */
-    public MemberAttribute getMemberAttribute() {
-        return memberAttribute;
+    public ConfigElementList<MemberAttribute> getMemberAttributes() {
+        return (memberAttribute == null) ? (memberAttribute = new ConfigElementList<MemberAttribute>()) : memberAttribute;
     }
 
     /**
@@ -60,7 +61,7 @@ public class GroupProperties extends ConfigElement {
      * @param memberAttribute the memberAttribute to set
      */
     @XmlElement(name = "memberAttribute")
-    public void setMemberAttribute(MemberAttribute memberAttribute) {
+    public void setMemberAttributes(ConfigElementList<MemberAttribute> memberAttribute) {
         this.memberAttribute = memberAttribute;
     }
 

